SPIKE.md

```diff
@@ -0,0 +1,126 @@
+
+## All Tests
+
+In order to run LoopBack 3's tests from their current folder, add LB3 tests'
+path to `test` entry in package.json:
+
+- `"test": "lb-mocha \"dist/**tests**/\*_/_.js\" \"lb3app/test/\*.js\""`
+
+In this case, the test folder is
+[`/lb3app/test`](https://github.com/strongloop/loopback-next/tree/spike/lb3test/examples/lb3-application/lb3app/test)
+from the root of the LoopBack 4 project.
+
+This will run LoopBack 4 tests first then LoopBack 3 tests.
+
+## Acceptance Tests
+
+First, move any LoopBack 3 test dependencies to package.json's devDependencies
+and run:
+
+```sh
+npm install
+```
+
+In your test file:
+
+- Update to use the LB4 Express server when doing requests:
+
+  ```ts
+  // can use lb4's testlab's supertest as the dependency is already installed
+  const request = require('@loopback/testlab').supertest;
+  const assert = require('assert');
+  const should = require('should');
+  const ExpressServer = require('../../dist/server').ExpressServer;
+
+  let app;
+
+  function json(verb, url) {
+    // use the LB4 express server
+    return request(app.server)
+      [verb](url)
+      .set('Content-Type', 'application/json')
+      .set('Accept', 'application/json')
+      .expect('Content-Type', /json/);
+  }
+  ```
+
+- Boot and start the LB4 app in your before hook, and stop the app in the after
+  hook:
+
+  ```ts
+  describe('LoopBack 3 style tests', function () {
+    before(async function () {
+      app = new ExpressServer();
+      await app.boot();
+      await app.start();
+    });
+
+    after(async () => {
+      await app.stop();
+    });
+
+    // your tests here
+  });
+  ```
+
+- Example of this use can be seen in
+  [`examples/lb3-application/lb3app/test/acceptance.js`](https://github.com/strongloop/loopback-next/blob/spike/lb3test/examples/lb3-application/lb3app/test/acceptance.js)
+  which has the same tests as
+  [`src/__tests__/acceptance/lb3app.acceptance.ts`](https://github.com/strongloop/loopback-next/blob/spike/lb3test/examples/lb3-application/src/__tests__/acceptance/lb3app.acceptance.ts),
+  but in LB3 style.
+
+Now when you run `npm test` your LoopBack 3 tests should be run along with any
+LoopBack 4 tests you have.
+
+Optional: Another option is to migrate your tests to use LoopBack 4 style of
+testing, similar to `src/__tests__/acceptance/lb3app.acceptance.ts`.
+Documentation for LoopBack testing can be found in
+https://loopback.io/doc/en/lb4/Testing-your-application.html.
+
+## Integration Tests
+
+For the integration tests, LoopBack 3 models were bound to the LoopBack 4
+application in order to allow JavaScript API to call application logic such as
+`Model.create()`. This can be seen in
+[`packages/booter-lb3app/src/lb3app.booter.ts`](https://github.com/strongloop/loopback-next/blob/spike/lb3test/packages/booter-lb3app/src/lb3app.booter.ts#L76-L85).
+
+In order to retrieve the model from the application's context,
+`getValueOrPromise()` can be used as follows:
```

Comment:
Please note that `getValueOrPromise` is an internal helper. Users should preferably use `get` (which returns a promise) or `getSync` (that throws an error if the value cannot be resolved synchronously).